controllers/cluster_controller.go

```diff
@@ -338,7 +372,14 @@ func splitMachineList(list *clusterv1.MachineList) (*clusterv1.MachineList, *clu
 	return controlplanes, nodes
 }
 
-func (r *ClusterReconciler) reconcileControlPlaneInitialized(ctx context.Context, cluster *clusterv1.Cluster) error {
+func (r *ClusterReconciler) reconcileControlPlaneInitialized(ctx context.Context, cluster *clusterv1.Cluster) (err error) {
+	defer func() {
```

Comment:
Instead of deferring in each single function, we should probably add another method to the list which is something like `reconcileMetrics` which runs after `reconcilePhase`.

Reply:
I'd prefer keeping them in the context of the reconcile functions because I find it to be more readable and easier to understand the metrics' calculation.
If we move it to a separate `reconcileMetrics` phase, it would be harder to unit test the metrics as part of each reconcile phase. Again, I think this would reduce the readability of the code.
There is also a chance that if other phases are added or updated, the `reconcileMetrics` which would be located elsewhere could be ignored thus leading to incorrect or missing metrics.

However, I do see the code repetition and the possibility that the defer func could grow as more metrics may be required in the future. WDYT.

Reply:
The main thing we should avoid is to use a method's intermediate state to evaluate metrics that seems to be generic, and very in line with something users get with a `kubectl describe`  call.

The concern around having the code somewhere else is understandable, although I don't think it's very different than what we have today.

Regarding tests, if the reconciliation for metrics is its own function, it'd much easier to test input/output/expectations by passing in specific scenarios to that function, rather than working on top of existing tests.

Reply:
Thanks for the feedback! 👍 
I understand it better now and I do agree about evaluating the state of the metrics at the end of the reconciliation for better consistency.